### PR TITLE
Balder/add coverage metrics

### PR DESCRIPTION
**Pull request by baldersheim**
@arnej27959 @jobergum @bratseth PR 
Note metric name "degraded_queries" and "coverage_per_query".
The former has the added dimension "reason" with 4 possible values
"match_phase", "timeout", "adaptive_timeout" and "non_ideal_state"

### CODE REVIEW

File: container-search/src/main/java/com/yahoo/prelude/statistics/StatisticsSearcher.java [hunk 1]
```diff
@@ -180,14 +170,43 @@ private void qps(Metric.Context metricContext) {
...
 
+    private Metric.Context getDegradedMetricContext(String chainName, Coverage coverage) {
+        Map<DegradedReason, Metric.Context> reasons = degradedReasonContexts.get(chainName);
+        if (reasons == null) {
+            DegradedReason [] reasonEnums = {
+                    DegradedReason.MatchPhase, DegradedReason.AdaptiveTimeout,
+                    DegradedReason.Timeout, DegradedReason.NonIdealState
+            };
+            reasons = new HashMap<>(4);
+            for (DegradedReason reason : reasonEnums ) {
```
Comment (bratseth) [inline container-search/src/main/java/com/yahoo/prelude/statistics/StatisticsSearcher.java]: Just use DegradedReason.values() bro ...

File: container-search/src/main/java/com/yahoo/prelude/statistics/StatisticsSearcher.java [hunk 2]
```diff
@@ -180,14 +170,43 @@ private void qps(Metric.Context metricContext) {
...
+                Map<String, String> dimensions = new HashMap<>();
+                dimensions.put("chain", chainName);
+                dimensions.put("reason", reason.toString());
+                Metric.Context context = this.metric.createContext(dimensions);
+                reasons.put(reason, context);
+            }
+            degradedReasonContexts.put(chainName, reasons);
+        }
+        if (coverage.isDegradedByMatchPhase()) {
```
Comment (bratseth) [inline container-search/src/main/java/com/yahoo/prelude/statistics/StatisticsSearcher.java]: I suggest you move DegradedReason to Coverage and create an accessor for it. Then you can replace these 10 lines by a single lookup.

**Review by bratseth — APPROVED** [review]
See my suggestions. Otherwise LGTM!

File: container-search/src/main/java/com/yahoo/prelude/statistics/StatisticsSearcher.java (hunk 1 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/prelude/statistics/StatisticsSearcher.java]: That was elegant, done 👍

File: container-search/src/main/java/com/yahoo/prelude/statistics/StatisticsSearcher.java (hunk 2 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/prelude/statistics/StatisticsSearcher.java]: That was my initial thought too. However as the reasons are not exclusive, and here we just select one of them, I concluded that it was better to have them here.
Ideally we should report all, but I think that will be more confusing.

**Pull request merged**